sabnzbd/dirscanner.py

```diff
@@ -124,90 +119,116 @@ def run(self):
+                            stat_tuple = os.stat(path) if sabnzbd.WIN32 else entry.stat()
                         except OSError:
                             continue
-                        if compare_stat_tuple(stat_tuple, stat_tuple_tmp):
-                            break
-                        stat_tuple = stat_tuple_tmp
                     else:
-                        # Not stable
+                        self.ignored[path] = 1
+                        yield path, catdir, None
                         continue
 
-                    # Add the NZB's
-                    res, _ = sabnzbd.nzbparser.add_nzbfile(path, catdir=catdir, keep=False)
-                    if res < 0:
-                        # Retry later, for example when we can't read the file
-                        self.suspected[path] = stat_tuple
-                    elif res == 0:
-                        self.error_reported = False
-                    else:
-                        self.ignored[path] = 1
+                    if path in self.suspected:
+                        if not compare_stat_tuple(self.suspected[path], stat_tuple):
+                            # Suspected file attributes have changed
+                            del self.suspected[path]
+
+                    yield path, catdir, stat_tuple
+        except OSError:
+            if not self.error_reported and not catdir:
+                logging.error(T("Cannot read Watched Folder %s"), filesystem.clip_path(folder))
+                self.error_reported = True
+
+    def __when_stable_add_nzbfile(self, path, catdir, stat_tuple):
+        """Try and import the NZB but wait until the attributes are stable for 1 second, but give up after 3 sec"""
+
+        logging.info("Trying to import %s", path)
+
+        # Wait until the attributes are stable for 1 second, but give up after 3 sec
+        # This indicates that the file is fully written to disk
+        for n in range(3):
+            with self.loop_condition:
+                self.loop_condition.wait(1.0)
+            if self.shutdown:
+                return
+
+            try:
+                stat_tuple_tmp = os.stat(path)
+            except OSError:
+                continue
+            if compare_stat_tuple(stat_tuple, stat_tuple_tmp):
+                break
+            stat_tuple = stat_tuple_tmp
+        else:
+            # Not stable
+            return
+
+        # Add the NZB's
+        res, _ = sabnzbd.nzbparser.add_nzbfile(path, catdir=catdir, keep=False)
+        if res < 0:
+            # Retry later, for example when we can't read the file
+            self.suspected[path] = stat_tuple
+        elif res == 0:
+            self.error_reported = False
+        else:
+            self.ignored[path] = 1
+
+    def scan(self):
+        """Do one scan of the watched folder"""
+
+        def run_dir(folder):
+            files = set()
+
+            with ThreadPoolExecutor() as executor:
```

Comment:
Does this not spawn new threads on every scan?
Is it worth it to just have a pool open and use that continuously? Or does it not really matter maybe?

Reply:
It won't create threads until jobs are submitted but does have the overhead of creating the executor instance each loop.

I think a much better solution would be to use asyncio, monitoring the paths until stable would be in a single thread and it's managed the executor or a custom restricted one calling the synchronous code which adds them to the queue.
Any objections if I put together that kind of approach? I think it would actually fit in quite nicely to when I add the event based part.

Reply:
Sounds good to try asyncio!